packages/flutter_tools/lib/src/cache.dart

```diff
@@ -95,8 +95,46 @@ class Cache {
   final Directory _rootOverride;
   final List<CachedArtifact> _artifacts = <CachedArtifact>[];
 
+  // Check whether there is a writable bit in the usr permissions.
+  static bool _hasUserWritePermission(FileStat stat) {
+    final int permissions = ((stat.mode & 0xFFF) >> 6) & 0x7;
+    return permissions == 2
+      || permissions == 3
+      || permissions == 6
+      || permissions == 7;
+  }
+
+  // Unfortunately the memory file system by default specifies a mode of `0`
+  // and is used by the majority of our tests. Default to false and only set
+  // to true when we know it is safe.
+  static bool checkPermissions = false;
+
   // Initialized by FlutterCommandRunner on startup.
-  static String flutterRoot;
+  static String get flutterRoot => _flutterRoot;
+  static String _flutterRoot;
+  static set flutterRoot(String value) {
+    if (value == null) {
+      _flutterRoot = null;
+      return;
+    }
+    if (checkPermissions) {
+      // Verify that we have writable permission in the flutter root. If not,
+      // we're liable to crash in unintuitive ways. This can happen if the user
+      // is using a homebrew or other unofficial channel, or otherwise installs
+      // Flutter into directory without permissions.
+      final FileStat binStat = fs.statSync(fs.path.join(value, 'bin'));
+      final FileStat rootStat = fs.statSync(value);
+      if (!_hasUserWritePermission(binStat) || !_hasUserWritePermission(rootStat)) {
+        throwToolExit(
+          'Warning: Flutter is missing permissions to write files '
+          'in its installation directory - "$value". '
```

Comment:
Perhaps we could hint the user that they should generally not install flutter in an area that requires Administrator/root permissions...

Reply:
I think that general it is more helpful to tell people what they should do instead of what they shouldn't

Reply:
So then just something like `Flutter should be installed in a location that does not require administrative or root access`?  It's the most general cause of the problem, and I've seen bugs about this e.g. when someone tries to install it in `C:\Program Files` on Windows.